### PR TITLE
Update to arrow 28

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -287,8 +287,8 @@ fn get_wider_decimal_type(
         (DataType::Decimal128(p1, s1), DataType::Decimal128(p2, s2)) => {
             // max(s1, s2) + max(p1-s1, p2-s2), max(s1, s2)
             let s = *s1.max(s2);
-            let range = (p1 - s1).max(p2 - s2);
-            Some(create_decimal_type(range + s, s))
+            let range = (*p1 as i8 - s1).max(*p2 as i8 - s2);
```
Comment (tustvold) [inline datafusion/expr/src/type_coercion/binary.rs:290]: Perhaps @liukun4515 you might be able to cast your eyes over this? My knowledge of decimals is fairly limited

File: datafusion/proto/proto/datafusion.proto [hunk 2]
```diff
@@ -705,8 +705,9 @@ enum IntervalUnit{
 }
 
 message Decimal{
-    uint64 whole = 1;
-    uint64 fractional = 2;
+  reserved 1, 2;
```
Comment (tustvold) [inline datafusion/proto/proto/datafusion.proto:708]: This is a breaking change to the protobuf serialization of schemas

File: datafusion/common/src/dfschema.rs [hunk 3]
```diff
@@ -947,12 +946,12 @@ mod tests {
     fn test_dfschema_to_schema_convertion() {
         let mut a: DFField = DFField::new(Some("table1"), "a", DataType::Int64, false);
         let mut b: DFField = DFField::new(Some("table1"), "b", DataType::Int64, false);
-        let mut a_metadata = BTreeMap::new();
```
Comment (alamb) [inline datafusion/common/src/dfschema.rs:950]: ❤️  -- thank you

File: datafusion/core/src/avro_to_arrow/schema.rs [hunk 4]
```diff
@@ -280,16 +280,6 @@ fn external_props(schema: &AvroSchema) -> BTreeMap<String, String> {
     props
 }
 
-#[allow(dead_code)]
```
Comment (alamb) [inline datafusion/core/src/avro_to_arrow/schema.rs:283]: 🧹  👍

**Review by alamb — COMMENTED** [review]
Thanks @tustvold  -- this looks great. We can probably remove the `crates.io` patch and get this ready PR ready for review

I agree it would be great if @liukun4515  could review the change to Decimal 🙏

**Review by alamb — APPROVED** [review]

File: datafusion/expr/src/type_coercion/binary.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:290]: thanks for the mention for me.
I think arrow ecosystem support negative scale, but in the datafusion we can support `scale>=0`.
We can add checker in the SQL level to ensure `scale >=0` in the datafusion.
cc @alamb

Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:290]: In the datafusion, we just support the decimal128, and should make sure the `precision>=1 and precision<=38` and `scale>=0 and scale<=precision`.

Do you have any thoughts for that? @tustvold

File: datafusion/optimizer/src/simplify_expressions/utils.rs [hunk 5]
```diff
@@ -108,8 +106,12 @@ pub fn is_one(s: &Expr) -> bool {
         | Expr::Literal(ScalarValue::UInt64(Some(1))) => true,
         Expr::Literal(ScalarValue::Float32(Some(v))) if *v == 1. => true,
         Expr::Literal(ScalarValue::Float64(Some(v))) if *v == 1. => true,
-        Expr::Literal(ScalarValue::Decimal128(Some(v), _p, _s)) => {
-            *_s < DECIMAL128_MAX_PRECISION && POWS_OF_TEN[*_s as usize] == *v
+        Expr::Literal(ScalarValue::Decimal128(Some(v), _p, s)) => {
+            *s >= 0
```
Comment (liukun4515) [inline datafusion/optimizer/src/simplify_expressions/utils.rs:110]: 👍

File: datafusion/sql/src/planner.rs [hunk 6]
```diff
@@ -3213,7 +3213,7 @@ mod tests {
             let sql = "SELECT CAST(10 AS DECIMAL(0))";
             let err = logical_plan(sql).expect_err("query should have failed");
             assert_eq!(
-                r##"Internal("Decimal(precision = 0, scale = 0) should satisty `0 < precision <= 38`, and `scale <= precision`.")"##,
+                r##"Internal("Decimal(precision = 0, scale = 0) should satisfy `-38 < precision <= 38`, and `scale <= precision`.")"##,
```
Comment (liukun4515) [inline datafusion/sql/src/planner.rs:3216]: I'am confused about the error message. 
The origin message is more reasonable for me, and the precision should be `precision>0 and precision<=38` in the datafusion.
cc @alamb @tustvold

Comment (liukun4515) [reply datafusion/sql/src/planner.rs:3216]: @HaoYang670 I remember we have seen this issue about `SELECT CAST(10 AS DECIMAL(0))` and use the error precision in the SQL

File: datafusion/sql/src/utils.rs [hunk 7]
```diff
@@ -522,9 +522,12 @@ pub(crate) fn make_decimal_type(
...
 
     // Arrow decimal is i128 meaning 38 maximum decimal digits
-    if precision == 0 || precision > DECIMAL128_MAX_PRECISION || scale > precision {
+    if precision == 0
+        || precision > DECIMAL128_MAX_PRECISION
+        || scale.unsigned_abs() > precision
+    {
         Err(DataFusionError::Internal(format!(
-            "Decimal(precision = {}, scale = {}) should satisty `0 < precision <= 38`, and `scale <= precision`.",
+            "Decimal(precision = {}, scale = {}) should satisfy `-38 < precision <= 38`, and `scale <= precision`.",
```
Comment (liukun4515) [inline datafusion/sql/src/utils.rs:530]: the same issue.

File: datafusion/sql/src/planner.rs (hunk 6 above)
Comment (HaoYang670) [reply datafusion/sql/src/planner.rs:3216]: The type of `precision` is `usize`, so it should be always `>=0`.

File: datafusion/sql/src/utils.rs (hunk 7 above)
Comment (liukun4515) [inline datafusion/sql/src/utils.rs:530]:
```suggestion
            "Decimal(precision = {}, scale = {}) should satisfy `0 < precision <= 38`, and `scale <= precision`.",
```

File: datafusion/sql/src/planner.rs (hunk 6 above)
Comment (liukun4515) [reply datafusion/sql/src/planner.rs:3216]: this is caused by https://github.com/apache/arrow-datafusion/pull/4400/files#r1035496657

File: datafusion/expr/src/type_coercion/binary.rs (hunk 1 above)
Comment (tustvold) [reply datafusion/expr/src/type_coercion/binary.rs:290]: > but in the datafusion we can support scale>=0

From an outsiders perspective, at least w.r.t decimals, I think breaking compatibility with the arrow specification in this way would likely be surprising to people. Is there a particular reason to not support negative scale?

Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:290]: Using the negative scale in the arrow ecosystem is ok for me.
But in the datafusion(SQL level system), it's better to make consistent with other SQL system, like Spark,PG,MySQL.
In other SQL level system, I have not seen the usage of negative scale.

In the PG
```
postgres=# create table test(c1 decimal(10,-1));
ERROR:  NUMERIC scale -1 must be between 0 and precision 10
LINE 1: create table test(c1 decimal(10,-1));
```

In the Spark:

```
spark-sql> create table test_d(c1 decimal(10,-1));
Error in query:
extraneous input '-' expecting INTEGER_VALUE(line 1, pos 34)

== SQL ==
create table test_d(c1 decimal(10,-1))
```

**Review by liukun4515 — APPROVED** [review]
LGTM
Thanks @tustvold

File: datafusion/expr/src/type_coercion/binary.rs (hunk 1 above)
Comment (tustvold) [reply datafusion/expr/src/type_coercion/binary.rs:290]: Spark definitely supports negative scales, it might not expose them in its SQL frontend though

**Pull request merged**